server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java

```diff
@@ -1108,4 +1038,160 @@ public static ClusterBlock createUUIDBasedBlock(ClusterBlock clusterBlock) {
             clusterBlock.levels()
         );
     }
+
+    public static class OpenIndicesExecutor implements ClusterStateTaskExecutor<OpenIndicesTask> {
+
+        private final AllocationService allocationService;
+        private final IndexMetadataVerifier indexMetadataVerifier;
+        private final IndicesService indicesService;
+        private final ShardLimitValidator shardLimitValidator;
+
+        public OpenIndicesExecutor(
+            AllocationService allocationService,
+            IndexMetadataVerifier indexMetadataVerifier,
+            IndicesService indicesService,
+            ShardLimitValidator shardLimitValidator
+        ) {
+            this.allocationService = allocationService;
+            this.indexMetadataVerifier = indexMetadataVerifier;
+            this.indicesService = indicesService;
+            this.shardLimitValidator = shardLimitValidator;
+        }
+
+        @Override
+        public ClusterTasksResult<OpenIndicesTask> execute(ClusterState currentState, List<OpenIndicesTask> tasks) throws Exception {
+            ClusterTasksResult.Builder<OpenIndicesTask> builder = ClusterTasksResult.builder();
+            ClusterState state = currentState;
+
+            try {
+                // build an in-order de-duplicated array of all the indices to open
+                final Set<Index> indicesToOpen = new LinkedHashSet<>(tasks.size());
+                for (OpenIndicesTask task : tasks) {
+                    for (Index index : task.request.indices()) {
+                        indicesToOpen.add(index);
+                    }
+                }
+                Index[] indices = indicesToOpen.toArray(new Index[0]);
+
+                // open them
+                state = openIndices(indices, state);
+
+                // do a final reroute
+                final String indicesAsString = Arrays.toString(indices);
```

Comment:
I think it's fine to just say `indices opened` here, no need to build this potentially-massive string just for the health-change log message.

Reply:
Makes sense, 8eab29473c2687b87f717e37d129ae732a5ad4bf